### DISCUSSION TITLE
CIP31 - Enhanced File Encoding Support

### DISCUSSION THREAD

**Discussion by jdogresorg** [CounterpartyXCP/CIPs#109] [CIPs]
Counterparty currently has an issue where `file` data is being stored in the counterparty database, and we are unable to move forward with updates (like P2WSH) which would allow much larger transactions to happen on Counterparty, until we address the issue with `files` in CP and how to get away from storing them in the database.

After chatting with a few difference CP community members about their ideas, I have put together a CIP which I believe addresses the current `file` issues in Counterparty, and will allow us to move forward with P2WSH and taproot support.

The core changes here are :
1. Allow users to specify content/mime type for files
2. Stop storing files in counterparty database
3. Update CP API to allow retrieving of `file` data direct from raw transaction/bitcoind

Below is a rough draft of the CIP I am working on... any feedback would be appreciated.

# Proposed CIP Draft

        CIP: 31
        Title: Enhanced File Encoding Support
        Author: Jeremy Johnson (J-Dog)
        Discussions-To: https://github.com/CounterpartyXCP/cips/discussions/109
        Status: Draft
        Type: Standards
        Created: 2023-09-07

# Abstract
Support encoding a `file` in a counterparty transaction 

# Motivation
Provide a way for users to associate a chunk of data or a file with a Counterparty issuance or broadcast transaction.

# Rationale
The popularity of blockchain `file` encoding protocols such as [`Inscriptions`](https://ordinals.com/inscriptions) and [`Bitcoin Stamps`](https://stampchain.io/) indicates users want to store `file` data on a highly-available and permanent medium. 

Counterparty currently stores `file` data in the internal database, which is adding bloat and slowing down queries.

This CIP has been created to standardize the way that Counterparty deals with a `file`, allow specifying a `mime-type` for a `file`, and to migrate `file` data out of the database.

# Definitions
- `file` - Any file type 
- `mime-type` - Any of the currently defined MIME / Media Types according to the [IANA Spec](https://www.iana.org/assignments/media-types/media-types.xhtml)

# Specification
The following modifications are needed to support this CIP.

## Issuances
- Add `data_type` param to `create_issuance` API (default: null)
- Add `data_type` field to `issuances` table
- Add new issuance data format with `data_type` field (id=22)
- Store `data_type` info in `data_type` field in `issuances` table
- Store `description` info in `description` field if `data_type` is `null` or `plain/text` (up to 1024 characters)

## Broadcasts
- Add `data_type` param to `create_broadcast` API (default: null)
- Add `data_type` field to `broadcasts` table
- Add new broadcast data format with `data_type` field (id=31)
- Store `data_type` info in `data_type` field in `broadcasts` table
- Store `text` info in `text` field if `data_type` is `null` or `plain/text` (up to 1024 characters)

## `get_{table}` API 
- Add `db_data` param (default: `true`)
- Set `limit` to `100` when `db_data` param is `false` (limits abuse on bitcoind)
- `get_broadcasts`
    - Return `text` field from raw tx data (bitcoind) when `db_data` param is `false`
- `get_issuances`
    - Return `description` field from raw tx data (bitcoind) when `db_data` param is `false`

# Copyright
This document is placed in the public domain.

**Comment by jdogresorg** [comment]
pull request submitted 

https://github.com/CounterpartyXCP/cips/pull/111

**Comment by jotapea** [comment]
@jdogresorg I have multiple comments:

- Good to see the use of new message ids, thus keeping backwards compatibility, BUT this is still not solving the reset issuance change (which I still believe should be removed).
- This CIP seems to imply that the description and text fields would now be limited to a length of 1024 characters? That should be more explicit (and explaining why that was the number chosen).
- I don't see any mention of how these changes affect the `messages` table, which is the main table that the rest of the protocol depends on.

People using the old message ids would still be able to use the protocol as they have been using it, so I don't see how this is solving much. Maybe a better approach is to just add a completely new message type that is only about a file, and keep it only in the `messages` table. Then, standardize a way of referencing this from the `issuances`, `broadcasts`, or any other relevant table.

Also, can you elaborate on:
> will allow us to move forward with P2WSH and taproot support

It sounds like this is a requirement for a previous CIP, so maybe this should be incorporated into that previous CIP? Or have this CIP explain better those specific needs.

**Reply by jdogresorg** [reply]
> This CIP seems to imply that the description and text fields would now be limited to a length of 1024 characters?

Users can still store as much data in the `description` and `text` fields as they want. Internally, counterparty-lib would just be limiting what is stored in the database to 1024 characters. If the user wants to get the full description, they can do so via setting the `db_data` param to false, which would then retrieve the full `description` or `text` field from the raw bitcoin transaction instead of the database.

>I don't see any mention of how these changes affect the messages table, which is the main table that the rest of the protocol depends on.

`messages` table data is generated from the data inserted into the CP database, so the changes listed in the CIP would also be reflected in the `messages` table (new `data_type` field and only storing up to 1024 characters in the database)

> Also, can you elaborate on "will allow us to move forward with P2WSH and taproot support"?

Counterparty should not support P2WSH and taproot which enable encoding of much larger files/data, until we implement a method to migrate `file` data out of the database going forward. Otherwise, the CP database could very rapidly explode in size and slow down considerably due to storing huge files in the database unnecessarily.

**Reply by jotapea** [reply]
NACK

This CIP fails to follow what I believe are core assumptions of the protocol, which is that everything is about the SQLite DB and starting at the `messages` table, not the other way around. The "explosion in DB size" is mainly a problem of data denormalization.

I can see this update (as proposed) going wrong in multiple aspects, complicating the elegance of being principally about confirming the state with SQL, and not really solving the core problems.

**Reply by Jpja** [reply]
>The "explosion in DB size" is mainly a problem of data denormalization.

@jotapea Do you think we have the resources to do a DB optimization (expertise, bounty funding, etc)?

If we cannot improve the DB, I'm leaning toward agreeing with @jdogresorg on limiting the asset description. ~1000 bytes seems about right.

Where I am not sold though, is on the need for adding so much complexity. I recommend making two simple rule changes
- description truncated at 1024 bytes (as in CIP 31; anything after not stored in the DB but can be retrieved from the BTC tx)
- ignore anything in a description after `##`

Meta rules for storing files (like STAMPS) can remain informational CIPs without protocol changes. STAMPS could benefit if they start storing store data chunks like `stamp:image/png;##{image_file_blob}` – 100% efficient versus base64's 6/8 today. Retrieving images is pretty straight forward. Bitcoin APIs with tools like https://jpja.github.io/Electrum-Counterparty/decode_tx.

**Reply by jotapea** [reply]
(I'm not going to answer the resources question, I won't go there. I'll just say that I am resource, and we need more developers reviewing code before merges to master.)

The proposed solution is just ugly. If it is implemented as proposed, I won't upgrade. I see it as a reactionary change, going deeper into hacks, instead of a thoughtful long term improvement.

Lets go into the specifics of what is a Counterparty transaction.

The Bitcoin transaction contains a CNTRPRTY message. This message is stored in the `transactions` table, as is. (So already I don't understand why this CIP specifies the "look into the raw bitcoind transaction" step when it is not necessary.) Then the message in the transaction, if supported and interpreted correctly by the protocol, creates 1+ other messages that all go into the `messages` table, after being inserted to SQL tables which help in the verifiability of it all.

That's it. This is the core of the protocol.

Arbitrary truncation of data will just complicate development on top of the protocol even more, and might even break it in terms of hash calculations as now there will be truncated data that is not considered (it can start similar to others, then change after the truncation).

xcp.dev, which was critical in showing that a contentious fork was not needed, will be broken by this change. A full application can be build on the Counterparty protocol by simply using the SQLite tables directly, this change will now require complicating the queries every time a text is longer than X.

counterparty-lib needs to do less, not more! The current read-api of the protocol is very inefficient, and this change will just make applications depend on it even more.

**Reply by jdogresorg** [reply]
>The Bitcoin transaction contains a CNTRPRTY message. This message is stored in the transactions table, as is. (So already I don't 
understand why this CIP specifies the "look into the raw bitcoind transaction" step when it is not necessary.) 

Currently, all transaction data is written to the `transactions` table... so, if counterparty was to enable taproot support and people can encode 500K+ size files into CP txs, now CP (as it is now) would not only include that full 500K of data in the `issuances` table, but also another 500k in the `messages` table, and yet another 500k in the `transactions` table.

Using the 500K file example, one could increase the size of the CP database 1.5 megs with a single transaction.

CIP31 only stores up to 1024 characters (1k) in the `messages`, `issuances`, and `transactions` table, limiting any database bloat to a max of 3k, while STILL allowing CP to retrieve the "full" asset description using a simple API flag. This keeps the database small/fast.

**Reply by jotapea** [reply]
> if counterparty was to enable taproot

That is not the current counterparty, and I am not sure how this is affecting [compatibility](https://github.com/CounterpartyXCP/cips/discussions/101#discussioncomment-6518789) for old bitcoin nodes.

From https://github.com/CounterpartyXCP/cips/discussions/109#discussioncomment-6981597:

> Maybe a better approach is to just add a completely new message type that is only about a file, and keep it only in the messages table.

We are both clear that everything in counterparty-lib's DB is copied multiple times. And we were talking just about the core tables.

Then, the `issuances` table continues copying data unnecessarily. And this is the one causing the most bloating, you agree?

**Reply by Jpja** [reply]
Some historical context. Asset descriptions and broadcasts were capped until sometime 2015. I don't think removing the limit was fully thought through, or at least they did not expect STAMPs back then. Re-enabling a limit won't be against any principles imo.

I agree that continued copying in the `issuances` table seems to me to be the most serious problem. For context, with LOCK and SWEEP the description is NOT embedded in the BTC tx, so long descriptions are duplicated for free. Isn't it planned that this will be the behavior for all issuances going forward?

Maybe we should make the change as simple as limiting asset descriptions at 1024 bytes? This and nothing else. Broadcasts can remain uncapped. Files can be stored as broadcasts.

**Reply by jotapea** [reply]
I would agree to just limiting descriptions... but this will mean losing STAMPS as part of Counterparty. The proposed hack "may allow" stamps to continue, but it does not solve the unnecessary copying of data in the `issuances` table, and pre-change descriptions will continue to be copied. And it is still an ugly hack that complicates the protocol for all future development.

Data normalization is a database design principle in which the resulting tables contain no duplicate data and still allow any query to be possible. It is the "ideal" state a database design can be... but in practice it might not. While high data normalization means reducing database size, the main issue with it is performance, as queries will have to do multiple joins to achieve results and these take more time to process.

There comes data denormalization, which means having some duplicated data for the sake of having better performing queries. But now this data copying will increase the database size... the tradeoffs of software.

The reason I am explaining this is to come back to the design of the `issuances` table. And I apologize for sounding like a broken record, but I have to bring up the `reset` again. Unfortunately, this change mostly cemented the issuances data design as it currently exists. Before the reset, it would have been possible to reduce the database size impact by improving normalization of the issuances table. It could have looked something like this:

From a single table (pre-reset implementation):
```sql
CREATE TABLE IF NOT EXISTS issuances(
  tx_index INTEGER PRIMARY KEY,
  tx_hash TEXT UNIQUE,
  block_index INTEGER,
  asset TEXT,
  quantity INTEGER,
  divisible BOOL,
  source TEXT,
  issuer TEXT,
  transfer BOOL, # not needed, is the same as source != issuer
  callable BOOL, # deprecated
  call_date INTEGER, # deprecated
  call_price REAL, # deprecated
  description TEXT,
  fee_paid INTEGER, # not sure about why this is here (will remove in next tables)
  locked BOOL,
  status TEXT,
  asset_longname TEXT # no need to repeat if is already in the assets table
```

To multiple tables, improving normalization (and taking the opportunity to clean up some columns, as explained in their comments):

The first one would only require row inserts in genesis and with issuer transfers:
```sql
CREATE TABLE IF NOT EXISTS issuances_issuer(
  tx_index INTEGER PRIMARY KEY,
  tx_hash TEXT UNIQUE,
  block_index INTEGER,
  asset TEXT,
  source TEXT,
  issuer TEXT,
  divisible BOOL, # only required for genesis
  asset_longname TEXT, # only required for genesis, if applies
  status TEXT
```

Then, the quantities and description updates could be on their own tables, only doing inserts when those specific values change:

```sql
CREATE TABLE IF NOT EXISTS issuances_quantity(
  tx_index INTEGER PRIMARY KEY,
  tx_hash TEXT UNIQUE,
  block_index INTEGER,
  asset TEXT,
  source TEXT,
  quantity INTEGER,
  locked BOOL, # after locked, no more inserts are needed
  status TEXT
```

```sql
CREATE TABLE IF NOT EXISTS issuances_description(
  tx_index INTEGER PRIMARY KEY,
  tx_hash TEXT UNIQUE,
  block_index INTEGER,
  asset TEXT,
  source TEXT,
  description TEXT,
  locked BOOL, # NEW FEATURE: after locked, no more inserts are needed
  status TEXT
```

This design would mean that transfers / sweeps will not copy descriptions. It would also mean, that if an issuance only wants to change the quantity or lock of an asset, it also doesn't have to repeat the description...

💡 *Maybe, the main problem is with the "expected" UI/UX, because the protocol could just NOT copy descriptions and make the applications be aware of this (xcp.dev is ready, as it shows all history and not only the latest description).*

With the `reset`, then I guess the only option would be to separate the `issuances_description`, but is evident how that change broke the cohesiveness of the issuances design for a minuscule use case.

Anyway, my suggestions are not the only option. If a "reset" like feature is still desired, then the divisibility can be moved to the `issuances_quantity` table and the genesis (name registration) issuance won't have to set divisibility. But this one would still repeat the divisible property every time (thus breaking normalization). Another idea could be that destroys now become negative `issuances_quantity` inserts...

In conclusion, there are many possible solutions and I'm more inclined into the core redesign for the long term than hacks that make future improvements even harder.

**Reply by Jpja** [reply]
> With the reset, then I guess the only option would be to separate the `issuances_description`.

Let us do this! If it's not too difficult? Solves the problem of duplicate descriptions. Clean and fair from a "gas" perspective. No need to limit the length then.

**Reply by jdogresorg** [reply]
> No need to limit the length then.

I fail to see how this removes the need to limit data length. Without limiting length of descriptions stored in database, we would still have the same issue of people being able to bloat database by storing files in them. 

@jotapea 's proposal is an interesting one that has some merit, but I don't feel that separating the issuances table into 3 separate tables at this time is the best solution.

**Reply by jotapea** [reply]
If descriptions are going to get limited, I would just remove them "entirely" (replaced with broadcasts).

Counterparty's core consensus, what is actually verified in every block, is all about the quantities of assets. If we want to make breaking changes for the long term then maybe this is the best solution. We can keep broadcasts as the only place for data needs.

The issuance command could even continue accepting descriptions, but behind the scenes, these are stored as broadcasts (obviously in the same transaction by the same issuer). We could add a "is_asset_description:<asset_name>" to these broadcasts.

This would allow STAMPS to continue, while at the same time solving all description copying by the DB.

**Reply by adamkrellenstein** [reply]
So I'm not fully onboarded as to all of the details and history here, but my instinct is to *de-normalize* our database in general, given that it's only 9GB (while Bitcoin + `addrindexrs` is 740GB!) Storage is cheap, and new blocks only happen every ~10 minutes (so there's plenty of time to process them), but we want to keep the protocol and the backend software flexible and responsive.

**Reply by jotapea** [reply]
Totally agree. I believe being able to sync fast is more critical. The protocol's main consensus responsibility IS verifying block / transactions state with SQL.

Many don't have nodes because they will have to rely on a bootstrap anyway, so why even bother. Nodes are to verify consensus. Nowadays, the `counterparty` part takes 10+ times more than bitcoin and addrindexrs.

There are many things to optimize. And I believe a great approach is to optimize looking from the hash backwards. What is required to generate this hash, and then optimize from here.

**Reply by adamkrellenstein** [reply]
Oooof, we need to speed up that verification...

**Comment by keyuno** [comment]
Why not consider implementing a system where XCP is burned above a certain length of description? This way, we would have limited description space paid only with BTC fees, and anything beyond that threshold would also require burning XCP.

**Reply by jotapea** [reply]
Broadcasts are "unlimited" in size and don't require any XCP. And invalid issuance entries, which don't consume the XCP, also get inserted into the DB. These are some reasons why, in general, I don't think requiring XCP is an effective way to prevent data storage in the DB.

**Reply by keyuno** [reply]
I am currently concerned about the proposal to impose a hard limit on the description field. Allowing for an unrestricted description length while proportionally increasing the difficulty and cost serves as a method to deter more significant modifications to the protocol, which always entails a risk of unintended consequences (hierarchy among assets based on whether they were created before or after such changes).

**Reply by jdogresorg** [reply]
>Why not consider implementing a system where XCP is burned above a certain length of description?

While this is an interesting proposal, I think that the BTC fee associated with encoding any large amount of data or file, is high enough to serve as an anti-spam mechanism (the feature XCP normally serves). The purpose of this proposal is to streamline uploading of multiple types of files and **LIMIT** the amount of data written to the database.

IMO moving to a "pay more XCP to write more to database" is good in some cases, like in the case of sweeps where a single small tx can result in a bunch of database records being created ( issue [#1220](https://github.com/CounterpartyXCP/counterparty-lib/issues/1220) ) .... However, i don't feel it is appropriate in this case, as this would enable users to write huge files to the CP database if they paid enough XCP. We do not want ANYONE to be able to write large files to the database, regardless of if they pay an XCP premium or not.

> I am currently concerned about the proposal to impose a hard limit on the description field. 

To clarify, there is no proposal to impose a hard limit on the description field, just a limit on what is stored in the database. Users can still encode any file or text string up to any size, the only limitation is that we will now only store up to 1024 characters in the `description` field in the database. 

- Users can still encode descriptions longer than 1024 characters
- Users can still get the full asset description via the `get_issuances` or `get_asset_info`

I am supportive of a per-issuance fee (like 0.01 XCP per issuance), but do not feel that requiring usage of XCP based on the size of the issuance is appropriate at this time. If someone wants to pay $120 to encode a STAMP into a bitcoin transaction, they should be able to do so and NOT have to pay any additional XCP fee.

**Reply by keyuno** [reply]
> However, i don't feel it is appropriate in this case, as this would enable users to write huge files to the CP database if they paid enough XCP. We do not want ANYONE to be able to write large files to the database, regardless of if they pay an XCP premium or not.

In the end, the critical factor lies in the quantity of XCP being burned. The larger the data, the greater the amount of XCP should be burned.

> To clarify, there is no proposal to impose a hard limit on the description field, just a limit on what is stored in the database. Users can still encode any file or text string up to any size, the only limitation is that we will now only store up to 1024 characters in the description field in the database.

This limitation is concerning because the proposal modifies the asset structure 'under the hood,' potentially impacting the value of assets. As mentioned earlier, it might introduce an unintended consequence—a hierarchy between assets created before and after the imposed limitation.

**Comment by jotapea** [comment]
PR for review: https://github.com/CounterpartyXCP/counterparty-lib/pull/1276

**Comment by Jpja** [comment]
To fix duplicate descriptions, can't we simply move the `description` column over to a new `issuances_descriptions` table?

The `issuances` table will have `description` replaced with `description_hash`, and `issuances_descriptions` will have two columns; `description_hash` (primary key) and `description`. 

It is necessary to something like this **before** enabling the [special null description](https://github.com/CounterpartyXCP/counterparty-lib/pull/1262) which duplicates the last description.

**Reply by jotapea** [reply]
`description_hash` ? Can you elaborate on what you mean by that?

**Reply by Jpja** [reply]
`description_hash` = sha256 hash of `description`. Can use hex-encoded string to be consistent with other hashes in the DB.

**Reply by jotapea** [reply]
I believe that to be consistent with the rest of the protocol, message tables must include the main ways to query:

```sql
tx_index INTEGER,
tx_hash TEXT,
block_index INTEGER,
source TEXT,
asset TEXT, -- if applies
```

If relations are going to use other kinds of keys then we are talking about changing the already consistent approach taken. Not saying is wrong, but it must be well thought out holistically.

**Reply by jdogresorg** [reply]
@Jpja @jotapea It sounds like your ideas on how to manage the `issuances` table by breaking it up into many different tables is drastically different than the way counterparty currently operates... This thread is for discussing my proposed CIP31.

CIP31 only discusses adding the `data_type` field to the `broadcasts` and `issuances` table.

CIP31 - Enhanced File Encoding Support
https://github.com/CounterpartyXCP/cips/blob/master/cip-0031.md

CIP31 Pull Request (PR)
https://github.com/CounterpartyXCP/counterparty-lib/pull/1276

I have followed the current dev guidelines, submitted a CIP, and a pull request with the relevant code changes for CIP31.

If you guys want a different database architecture for the `issuances` table, I suggest that you put forth your own CIPs which include rationale and specifications clearly defining your ideas. Also please code up a PR to put your idea into actual code so that the other devs can review the actual code being implemented... that is the preferred / defined method to push your ideas forward.

**Reply by Jpja** [reply]
I opened an issue specifically for this.
https://github.com/CounterpartyXCP/counterparty-lib/issues/1287
